filters/f_autoconvert.c

```diff
@@ -195,26 +195,27 @@ static bool build_image_converter(struct mp_autoconvert *c, struct mp_log *log,
     bool hw_to_sw = !imgfmt_is_sw && dst_have_sw;
 
     if (dst_all_hw && num_fmts > 0) {
-        // We can probably use this! Very lazy and very approximate.
-        struct mp_hwupload *upload = mp_hwupload_create(conv, fmts[0]);
-        if (upload) {
-            int sw_fmt = imgfmt_is_sw ? img->imgfmt : img->params.hw_subfmt;
-
-            mp_info(log, "HW-uploading to %s\n", mp_imgfmt_to_name(fmts[0]));
-            filters[2] = upload->f;
-            hwupload_fmt = mp_hwupload_find_upload_format(upload, sw_fmt);
-            fmts = &hwupload_fmt;
-            num_fmts = hwupload_fmt ? 1 : 0;
-            hw_to_sw = false;
-
-            // We cannot do format conversions when transferring between
-            // two hardware devices, so fail immediately if that would be
-            // required.
-            if (!imgfmt_is_sw && hwupload_fmt != sw_fmt) {
-                mp_err(log, "Format %s is not supported by %s\n",
-                       mp_imgfmt_to_name(sw_fmt),
-                       mp_imgfmt_to_name(p->imgfmts[0]));
-                goto fail;
+        for (int i = 0; i < num_fmts; i++) {
+            // We can probably use this! Very lazy and very approximate.
+            struct mp_hwupload *upload = mp_hwupload_create(conv, fmts[i]);
+            if (upload) {
+                int sw_fmt = imgfmt_is_sw ? img->imgfmt : img->params.hw_subfmt;
+
+                mp_info(log, "HW-uploading to %s\n", mp_imgfmt_to_name(fmts[i]));
+                filters[2] = upload->f;
+                hwupload_fmt = mp_hwupload_find_upload_format(upload, sw_fmt);
+                fmts = &hwupload_fmt;
+                num_fmts = hwupload_fmt ? 1 : 0;
+                hw_to_sw = false;
+
+                // We cannot do format conversions when transferring between
+                // two hardware devices, so reject this format if that would be
+                // required.
+                if (!imgfmt_is_sw && hwupload_fmt != sw_fmt) {
+                    mp_err(log, "Format %s is not supported by %s\n",
+                           mp_imgfmt_to_name(sw_fmt),
+                           mp_imgfmt_to_name(p->imgfmts[i]));
+                }
```

Comment:
What happens if all the fmts are not supported? It will not `goto fail`.

Reply:
Looking at the code here, I'm actually confused at why it doesn't _already_ work like this. Currently it's just hardcoded to try only the very first fmt instead of looping through all of them. But yeah the complete failure condition should still be tracked. Something like:
```
diff --git a/filters/f_autoconvert.c b/filters/f_autoconvert.c
index b55673b448..4dd55f9bd8 100644
--- a/filters/f_autoconvert.c
+++ b/filters/f_autoconvert.c
@@ -193,6 +193,7 @@ static bool build_image_converter(struct mp_autoconvert *c, struct mp_log *log,
 
     // Source is hw, some targets are sw -> try to download.
     bool hw_to_sw = !imgfmt_is_sw && dst_have_sw;
+    bool upload_success = false;
 
     if (dst_all_hw && num_fmts > 0) {
         for (int i = 0; i < num_fmts; i++) {
@@ -215,9 +216,14 @@ static bool build_image_converter(struct mp_autoconvert *c, struct mp_log *log,
                     mp_err(log, "Format %s is not supported by %s\n",
                            mp_imgfmt_to_name(sw_fmt),
                            mp_imgfmt_to_name(p->imgfmts[i]));
+                } else {
+                    upload_success = true;
+                    break;
                 }
             }
         }
+        if (!upload_success)
+            goto fail;
     }
 
     int src_fmt = img->imgfmt;
```